python/lsst/daf/butler/datastores/posixDatastore.py

```diff
@@ -44,66 +69,115 @@ class PosixDatastore(Datastore):
-            self.storageClassFactory.registerStorageClass(newStorageClass)
+        # Somewhere to temporarily store dataset to formatter maps
+        self.internalRegistry = {}
 
-            # Create the formatter, indexed by the storage class
-            # Currently, we allow this to be optional because some storage classes
-            # are not yet defined fully.
-            if "formatter" in info:
-                self.formatterFactory.registerFormatter(name, info["formatter"])
+    def addStoredFileInfo(self, ref, info):
+        """Record formatter information associated with this `DatasetRef`
 
-        # Read the file naming templates
-        self.templates = FileTemplates(self.config["templates"])
+        Parameters
+        ----------
+        ref : `DatasetRef`
+            The Dataset that has been stored.
+        info : `StoredFileInfo`
+            Metadata associated with the stored Dataset.
 
-    def get(self, uri, storageClass, parameters=None):
-        """Load an `InMemoryDataset` from the store.
+        Notes
+        -----
+        Need to convert this to a real registry.
+        """
+        self.internalRegistry[ref.id] = (info.formatter, info.path, info.storageClass.name)
+
+    def getStoredFileInfo(self, ref):
+        """Retrieve information associated with file stored in this
+        `Datastore`.
 
         Parameters
         ----------
-        uri : `str`
-            a Universal Resource Identifier that specifies the location of the
-            stored `Dataset`.
-        storageClass : `StorageClass`
-            the `StorageClass` associated with the `DatasetType`.
+        ref : `DatasetRef`
+            The Dataset that is to be queried.
+
+        Returns
+        -------
+        info : `StoredFileInfo`
+            Stored information about the internal location of this file
+            and its formatter.
+
+        Notes
+        -----
+        Need to convert this to a real registry.
+        """
+        if ref.id not in self.internalRegistry:
+            raise KeyError("Unable to retrieve formatter associated with Dataset {}".format(ref.id))
+        formatter_name, path, storageClass = self.internalRegistry[ref.id]
+        # Convert name of StorageClass to instance
+        storageClass = self.storageClassFactory.getStorageClass(storageClass)
+        return StoredFileInfo(formatter_name, path, storageClass)
+
+    def exists(self, ref):
+        """Check if the dataset exists in the datastore.
+
+        Parameters
+        ----------
+        ref : `DatasetRef`
+            Reference to the required dataset.
+
+        Returns
+        -------
+        exists : `bool`
+            `True` if the entity exists in the `Datastore`.
+        """
+        # Get the file information (this will fail if no file)
+        try:
+            storedFileInfo = self.getStoredFileInfo(ref)
+        except KeyError:
+            return False
```

Comment:
Good point. What does existence mean...? I was treating existence as "can I retrieve this thing" not "has it ever existed".

Reply:
this is also a datastore.exists and not a butler.exists. If butler splits up an exposure and writes it as separate items datastore never sees the parent DatasetRef so can't understand that it does indeed have all the bits needed to make the composite. The only way this would really work is if butler could register the parent DatasetRef with the datastore after it's put all the components.

Reply:
Keeping Datastore from knowing about these parents is probably a good idea.  I think that just means we need a way for Registry to know that no Datastore entry *should* exist.  That was going to be the `assembler` field in Dataset, and I think we can now just make that into a bool.